// react-router-loading/dist/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findMatchingElement = exports.isLoadable = exports.isRoutesDifferent = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactRouter = require("react-router");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var isRoutesDifferent = function isRoutesDifferent(current, next) {
  var _next$location, _current$location;

  return ((_next$location = next.location) === null || _next$location === void 0 ? void 0 : _next$location.pathname) !== ((_current$location = current.location) === null || _current$location === void 0 ? void 0 : _current$location.pathname);
};

exports.isRoutesDifferent = isRoutesDifferent;

var findMatchingRoute = function findMatchingRoute(location, routes) {
  var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function (matchingElement) {};
  var match; // We use React.Children.forEach instead of React.Children.toArray().find()
  // here because toArray adds keys to all child elements and we do not want
  // to trigger an unmount/remount for two <Route>s that render the same
  // component at different URLs.

  _react["default"].Children.forEach(routes, function (child) {
    if (match == null && /*#__PURE__*/_react["default"].isValidElement(child)) {
      var path = child.props.path || child.props.from;
      match = (0, _reactRouter.matchPath)(location === null || location === void 0 ? void 0 : location.pathname, _objectSpread(_objectSpread({}, child.props), {}, {
        path: path
      }));
      if (match) callback(child);
    }
  });

  return match;
};

var isLoadable = function isLoadable(location, routes) {
  var isLoadable;
  var match = findMatchingRoute(location, routes, function (matchingElement) {
    isLoadable = matchingElement.props.loading;
  });
  return match ? isLoadable : false;
};

exports.isLoadable = isLoadable;

var findMatchingElement = function findMatchingElement(location, routes) {
  var element;
  var match = findMatchingRoute(location, routes, function (matchingElement) {
    element = matchingElement;
  });
  return match ? /*#__PURE__*/_react["default"].cloneElement(element, {
    location: location,
    computedMatch: match
  }) : null;
};

exports.findMatchingElement = findMatchingElement;
